### DISCUSSION TITLE
Checkpointing support

### DISCUSSION THREAD

**Discussion by sdrave** [pymor/pymor#1192] [Ideas]
@HenKlei added a SIGINT handler in #1188 which will write the current reduced model to disk when training is interrupted by the user by pressing `CTRL-C` or otherwise sending SIGINT to the process. This will allow the user to work with the current reduced model or continue training at a different point in time.

I am wondering whether there is interest for such a feature elsewhere in pyMOR (`greedy` algorithms obviously come to mind). Similar but not exactly the same, some form of [checkpointing](https://en.wikipedia.org/wiki/Checkpointing) might be useful feature. So, who is interested in such features and for what?

**Comment by pmli** [comment]
On the sys-mor side, there are also a couple of iterative methods, e.g., IRKA and LR-ADI, that could have this. It's probably not good for reproducibility, but it might be nice to have when experimenting. Typically, if an iteration is taking a long time, I would cancel it and restart it with a smaller maximum number of iterations and maybe some different options, and this works fine for me, so I don't feel strongly about adding checkpointing support.

**Comment by renefritze** [comment]
Timesteppers would also be good candidate for checkpointing I think. No idea how the python ecosystem is for this, but in the HPC world there exist dedicated libraries  for checkpointing to allow minimal impact on the main computation. Stuff like async I/O and only recording deltas are at least things we should also consider.

**Reply by sdrave** [reply]
Have you seen anything that works well with Python? I only looked very briefly and didn't find anything promising.

**Reply by renefritze** [reply]
No, as I said, I've no idea how the situation in Python is. I also haven't looked though.

**Comment by HenKlei** [comment]
After #1188 was removed from the current release, I wanted to bring this topic up once again.

If there is serious interest in checkpointing support, I would take a look into the options we have in Python.

To shortly summarize (correct me please, if I misunderstood anything): From the sys-mor side, @pmli does not see a strong necessity for this. From the RB side, greedy algorithms, time steppers and the neural network training could benefit from such features.
